Migration Error from 55.11 to 56.2

### Describe the bug

Not everyone will hit this but a specific customer did and we don't know exactly why. Seems this happened again:

https://github.com/metabase/metabase/issues/42711

### To Reproduce

All the context is here:

https://metaboat.slack.com/archives/C05MPF0TM3L/p1755094785844409


### Expected behavior

Upgrade works

### Logs

2025-08-13 14:15:47.858	
{"exception":{"exception_class":"java.sql.BatchUpdateException","exception_message":"Batch entry 0 UPDATE databasechangelog SET FILENAME = CASE WHEN ID = ('v00.00-000') THEN ('migrations/001_update_migrations.yaml') WHEN ID < ('v45.00-001') THEN ('migrations/000_legacy_migrations.yaml') ELSE ('migrations/001_update_migrations.yaml') END was aborted: ERROR: duplicate key value violates unique constraint \"idx_databasechangelog_id_author_filename\"\n  Detail: Key (id, author, filename)=(v56.2025-06-05T16:48:48, edpaget, migrations/001_update_migrations.yaml) already exists.  Call getNextException to see other errors in the batch.","stacktrace":"java.sql.BatchUpdateException: Batch entry 0 UPDATE databasechangelog SET FILENAME = CASE WHEN ID = ('v00.00-000') THEN ('migrations/001_update_migrations.yaml') WHEN ID < ('v45.00-001') THEN ('migrations/000_legacy_migrations.yaml') ELSE ('migrations/001_update_migrations.yaml') END was aborted: ERROR: duplicate key value violates unique constraint \"idx_databasechangelog_id_author_filename\"\n  Detail: Key (id, author, filename)=(v56.2025-06-05T16:48:48, edpaget, migrations/001_update_migrations.yaml) already exists.  Call getNextException to see other errors in the batch.\n\tat org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:165)\n\tat org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2422)\n\tat org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:580)\n\tat org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:886)\n\tat org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:910)\n\tat org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1778)\n\tat com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeBatch(NewProxyPreparedStatement.java:2399)\n\tat clojure.java.jdbc$execute_batch.invokeStatic(jdbc.clj:598)\n\tat clojure.java.jdbc$execute_batch.invoke(jdbc.clj:591)\n\tat clojure.java.jdbc$db_do_execute_prepared_statement$fn__38758.invoke(jdbc.clj:1057)\n\tat clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:807)\n\tat clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)\n\tat clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:789)\n\tat clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)\n\tat clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:1056)\n\tat clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:1042)\n\tat clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:1080)\n\tat clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:1060)\n\tat clojure.java.jdbc$execute_BANG_$execute_helper__38826.invoke(jdbc.clj:1464)\n\tat clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1466)\n\tat clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1435)\n\tat clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1456)\n\tat clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1435)\n\tat metabase.app_db.liquibase$consolidate_liquibase_changesets_BANG_$fn__46226.invoke(liquibase.clj:473)\n\tat metabase.app_db.liquibase$run_in_scope_locked$reify__46148.run(liquibase.clj:344)\n\tat liquibase.Scope.lambda$child$0(Scope.java:216)\n\tat liquibase.Scope.child(Scope.java:225)\n\tat liquibase.Scope.child(Scope.java:215)\n\tat liquibase.Scope.child(Scope.java:194)\n\tat metabase.app_db.liquibase$run_in_scope_locked.invokeStatic(liquibase.clj:337)\n\tat metabase.app_db.liquibase$run_in_scope_locked.invoke(liquibase.clj:320)\n\tat metabase.app_db.liquibase$consolidate_liquibase_changesets_BANG_.invokeStatic(liquibase.clj:472)\n\tat metabase.app_db.liquibase$consolidate_liquibase_changesets_BANG_.invoke(liquibase.clj:451)\n\tat metabase.app_db.setup$migrate_BANG_$fn__46269.invoke(setup.clj:85)\n\tat metabase.app_db.liquibase$do_with_liquibase$f_STAR___46084.invoke(liquibase.clj:139)\n\tat metabase.app_db.liquibase$do_with_liquibase.invokeStatic(liquibase.clj:142)\n\tat metabase.app_db.liquibase$do_with_liquibase.invoke(liquibase.clj:130)\n\tat metabase.app_db.setup$migrate_BANG_.invokeStatic(setup.clj:78)\n\tat metabase.app_db.setup$migrate_BANG_.doInvoke(setup.clj:59)\n\tat clojure.lang.RestFn.invoke(RestFn.java:428)\n\tat metabase.app_db.setup$run_schema_migrations_BANG_.invokeStatic(setup.clj:186)\n\tat metabase.app_db.setup$run_schema_migrations_BANG_.invoke(setup.clj:181)\n\tat metabase.app_db.setup$setup_db_BANG_$fn__46324$fn__46325.invoke(setup.clj:204)\n\tat metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:239)\n\tat metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:225)\n\tat metabase.app_db.setup$setup_db_BANG_$fn__46324.invoke(setup.clj:198)\n\tat metabase.app_db.setup$setup_db_BANG_.invokeStatic(setup.clj:197)\n\tat metabase.app_db.setup$setup_db_BANG_.invoke(setup.clj:190)\n\tat metabase.app_db.core$setup_db_BANG_$fn__46519.invoke(core.clj:122)\n\tat metabase.app_db.core$setup_db_BANG_.invokeStatic(core.clj:117)\n\tat metabase.app_db.core$setup_db_BANG_.doInvoke(core.clj:104)\n\tat clojure.lang.RestFn.invoke(RestFn.java:424)\n\tat metabase.core.core$init_BANG__STAR_.invokeStatic(core.clj:138)\n\tat metabase.core.core$init_BANG__STAR_.invoke(core.clj:117)\n\tat metabase.core.core$init_BANG_.invokeStatic(core.clj:194)\n\tat metabase.core.core$init_BANG_.invoke(core.clj:189)\n\tat metabase.core.core$start_normally.invokeStatic(core.clj:208)\n\tat metabase.core.core$start_normally.invoke(core.clj:200)\n\tat metabase.core.core$entrypoint.invokeStatic(core.clj:243)\n\tat metabase.core.core$entrypoint.doInvoke(core.clj:234)\n\tat clojure.lang.RestFn.invoke(RestFn.java:400)\n\tat clojure.lang.AFn.applyToHelper(AFn.java:152)\n\tat clojure.lang.RestFn.applyTo(RestFn.java:135)\n\tat clojure.lang.Var.applyTo(Var.java:707)\n\tat clojure.core$apply.invokeStatic(core.clj:667)\n\tat clojure.core$apply.invoke(core.clj:662)\n\tat metabase.core.bootstrap$_main.invokeStatic(bootstrap.clj:36)\n\tat metabase.core.bootstrap$_main.doInvoke(bootstrap.clj:29)\n\tat clojure.lang.RestFn.invoke(RestFn.java:400)\n\tat clojure.lang.AFn.applyToHelper(AFn.java:152)\n\tat clojure.lang.RestFn.applyTo(RestFn.java:135)\n\tat metabase.core.bootstrap.main(Unknown Source)\nCaused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint \"idx_databasechangelog_id_author_filename\"\n  Detail: Key (id, author, filename)=(v56.2025-06-05T16:48:48, edpaget, migrations/001_update_migrations.yaml) already exists.\n\tat org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2734)\n\tat org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2421)\n\t... 70 more\n"},"lvl":"ERROR","m":"Metabase Initialization FAILED","lgr":"metabase.core.core"}

### Information about your Metabase installation

```JSON
56.2
```

### Severity

severe for who upgrades

### Additional context

_No response_

--- a/src/metabase/app_db/liquibase.clj
+++ b/src/metabase/app_db/liquibase.clj
@@ -478,10 +478,22 @@
             "v45.00-001" legacy-migrations-file
             "v56.0000-00-00T00:00:00" update001-migrations-file]))))))
 
+(def ^:private special-case-migrations #{"v56.2025-06-05T16:48:48" "v56.2025-05-19T16:48:48"})
+
+(defn- handle-special-case-migrations
+  "This handles v56 migrations that were checked into the v55 branch to resolve an issue with
+  inadventently backported migrations in 55. We check if this or the bad backports are the most recent
+  available migration and explicitly return 55 as the available major version if so."
+  [s]
+  (when (contains? special-case-migrations s)
+    55))
+
 (defn- extract-numbers
   "Returns contiguous integers parsed from string s"
   [s]
-  (map #(Integer/parseInt %) (re-seq #"\d+" s)))
+  (if-let [special-cased (handle-special-case-migrations s)]
+    [special-cased]
+    (map #(Integer/parseInt %) (re-seq #"\d+" s))))
 
 (defn latest-available-major-version
   "Get the latest version that Liquibase would apply if we ran migrations right now."

--- a/test/metabase/app_db/liquibase_test.clj
+++ b/test/metabase/app_db/liquibase_test.clj
@@ -153,6 +153,11 @@
         (.update liquibase "")
         (is (< 52 (liquibase/latest-applied-major-version conn (.getDatabase liquibase))))))))
 
+(deftest extract-numbers-special-case-test
+  (testing "when specific migration verison is passed reports different major version"
+    (is (= 55 (first (#'liquibase/extract-numbers "v56.2025-06-05T16:48:48"))))
+    (is (= 55 (first (#'liquibase/extract-numbers "v56.2025-05-19T16:48:48"))))))
+
 (deftest rollback-major-version
   (mt/test-drivers #{:h2 :mysql :rollback}
     (mt/with-temp-empty-app-db [conn driver/*driver*]

